Flaky `GrizzlyThriftClientTest`

https://github.com/eclipse-ee4j/glassfish-grizzly-thrift/actions/runs/33640056544/job/100280575439?pr=33#step:4:3513

https://github.com/eclipse-ee4j/glassfish-grizzly-thrift/actions/runs/33640781858/job/100283008905?pr=35#step:4:3520

```
[ERROR] org.glassfish.grizzly.thrift.client.GrizzlyThriftClientTest.testBasic  Time elapsed: 0.015 s  <<< ERROR!
java.net.BindException: Address already in use
	at org.glassfish.grizzly.thrift.client.GrizzlyThriftClientTest.createThriftServer(GrizzlyThriftClientTest.java:317)
	at org.glassfish.grizzly.thrift.client.GrizzlyThriftClientTest.testBasic(GrizzlyThriftClientTest.java:61)
```

--- a/src/test/java/org/glassfish/grizzly/thrift/TestUtils.java
+++ b/src/test/java/org/glassfish/grizzly/thrift/TestUtils.java
@@ -1,0 +1,56 @@
+/*
+ * Copyright (c) 2026 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+package org.glassfish.grizzly.thrift;
+
+import java.time.Duration;
+import java.util.concurrent.Callable;
+
+public class TestUtils {
+
+    private TestUtils() {
+        // Prevent instantiation
+    }
+
+    /**
+     * Repeatedly attempts to execute the given action. If the action throws an exception or assertion error,
+     * it pauses and retries until the timeout is reached.
+     *
+     * @param action        the task to execute
+     * @param timeout the maximum time to wait in milliseconds
+     * @param sleep   the pause duration between retries in milliseconds
+     * @param <T>           the type of the result
+     * @return the result of the action if it succeeds before the timeout
+     * @throws Exception if the timeout is exceeded, the last thrown exception is rethrown
+     */
+    @SuppressWarnings("SleepWhileInLoop")
+    public static <T> T retryUntilSuccess(Callable<T> action, Duration timeout, Duration sleep) throws Exception {
+        final long startTime = System.currentTimeMillis();
+        Exception lastException = null;
+        final long timeoutMillis = timeout.toMillis();
+        final long sleepMillis = sleep.toMillis();
+
+        while (System.currentTimeMillis() - startTime <= timeoutMillis) {
+            try {
+                return action.call();
+            } catch (Exception t) {
+                lastException = t;
+            }
+            Thread.sleep(sleepMillis);
+        }
+
+        throw lastException;
+    }
+}

--- a/src/test/java/org/glassfish/grizzly/thrift/ThriftTutorialTest.java
+++ b/src/test/java/org/glassfish/grizzly/thrift/ThriftTutorialTest.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright (c) 2026 Contributors to the Eclipse Foundation.
  * Copyright (c) 2011, 2017 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
@@ -60,6 +61,7 @@ import org.junit.runners.Parameterized;
 import org.junit.runners.Parameterized.Parameters;
 import org.junit.runner.RunWith;
 
+import static java.time.Duration.ofSeconds;
 import static org.junit.Assert.*;
 
 /**
@@ -116,7 +118,7 @@ public class ThriftTutorialTest {
         transport.setProcessor(filterChainBuilder.build());
 
         try {
-            transport.bind(PORT);
+            TestUtils.retryUntilSuccess(() -> transport.bind(PORT), ofSeconds(10), ofSeconds(1));
             transport.start();
 
             for (int i = 0; i < clientsNum; i++) {

--- a/src/test/java/org/glassfish/grizzly/thrift/client/GrizzlyThriftClientFailoverTest.java
+++ b/src/test/java/org/glassfish/grizzly/thrift/client/GrizzlyThriftClientFailoverTest.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright (c) 2026 Contributors to the Eclipse Foundation.
  * Copyright (c) 2014, 2017 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
@@ -24,6 +25,7 @@ import org.glassfish.grizzly.filterchain.TransportFilter;
 import org.glassfish.grizzly.nio.transport.TCPNIOTransport;
 import org.glassfish.grizzly.nio.transport.TCPNIOTransportBuilder;
 import org.glassfish.grizzly.thrift.CalculatorHandler;
+import org.glassfish.grizzly.thrift.TestUtils;
 import org.glassfish.grizzly.thrift.ThriftFrameFilter;
 import org.glassfish.grizzly.thrift.ThriftServerFilter;
 import org.junit.Assert;
@@ -35,6 +37,7 @@ import java.net.InetSocketAddress;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicInteger;
 
+import static java.time.Duration.ofSeconds;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
@@ -45,9 +48,9 @@ import static org.junit.Assert.assertTrue;
  */
 public class GrizzlyThriftClientFailoverTest {
 
-    private static final int PORT = 7791;
-    private static final int INVALID_PORT1 = 7800;
-    private static final int FAILOVER_PORT = 7792;
+    private static final int PORT = 7795;
+    private static final int INVALID_PORT1 = 7796;
+    private static final int FAILOVER_PORT = 7797;
 
     @Test
     public void testServerRemains() throws Exception {
@@ -302,14 +305,14 @@ public class GrizzlyThriftClientFailoverTest {
         }
     }
 
-    private static TCPNIOTransport createThriftServer(final int port, final ThriftServerFilter filter) throws IOException {
+    private static TCPNIOTransport createThriftServer(final int port, final ThriftServerFilter filter) throws Exception {
         final FilterChainBuilder filterChainBuilder = FilterChainBuilder.stateless();
         filterChainBuilder.add(new TransportFilter());
         filterChainBuilder.add(new ThriftFrameFilter());
         filterChainBuilder.add(filter);
         final TCPNIOTransport transport = TCPNIOTransportBuilder.newInstance().build();
         transport.setProcessor(filterChainBuilder.build());
-        transport.bind(port);
+        TestUtils.retryUntilSuccess(() -> transport.bind(port), ofSeconds(10), ofSeconds(1));
         transport.start();
         return transport;
     }

--- a/src/test/java/org/glassfish/grizzly/thrift/client/GrizzlyThriftClientTest.java
+++ b/src/test/java/org/glassfish/grizzly/thrift/client/GrizzlyThriftClientTest.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright (c) 2026 Contributors to the Eclipse Foundation.
  * Copyright (c) 2012, 2017 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
@@ -23,6 +24,7 @@ import org.glassfish.grizzly.http.server.NetworkListener;
 import org.glassfish.grizzly.nio.transport.TCPNIOTransport;
 import org.glassfish.grizzly.nio.transport.TCPNIOTransportBuilder;
 import org.glassfish.grizzly.thrift.CalculatorHandler;
+import org.glassfish.grizzly.thrift.TestUtils;
 import org.glassfish.grizzly.thrift.ThriftFrameFilter;
 import org.glassfish.grizzly.thrift.ThriftServerFilter;
 import org.glassfish.grizzly.thrift.http.ThriftHttpHandler;
@@ -40,6 +42,7 @@ import java.util.HashSet;
 import java.util.Set;
 import java.util.concurrent.CountDownLatch;
 
+import static java.time.Duration.ofSeconds;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
@@ -307,14 +310,14 @@ public class GrizzlyThriftClientTest {
         thriftServer2.shutdownNow();
     }
 
-    private static TCPNIOTransport createThriftServer(final int port, final Calculator.Processor tprocessor) throws IOException {
+    private static TCPNIOTransport createThriftServer(final int port, final Calculator.Processor tprocessor) throws Exception {
         final FilterChainBuilder filterChainBuilder = FilterChainBuilder.stateless();
         filterChainBuilder.add(new TransportFilter());
         filterChainBuilder.add(new ThriftFrameFilter());
         filterChainBuilder.add(new ThriftServerFilter(tprocessor));
         final TCPNIOTransport transport = TCPNIOTransportBuilder.newInstance().build();
         transport.setProcessor(filterChainBuilder.build());
-        transport.bind(port);
+        TestUtils.retryUntilSuccess(() -> transport.bind(port), ofSeconds(10), ofSeconds(1));
         transport.start();
         return transport;
     }

